tokio/src/sync/semaphore.rs

```diff
@@ -102,6 +102,78 @@ use std::sync::Arc;
 /// }
 /// ```
 ///
+/// Implement a simple token bucket for rate limiting
+///
+/// Many applications and systems have constraints on the rate at which certain
+/// operations should occur. Exceeding this rate can result in suboptimal
+/// performance or even errors.
+///
+/// The provided example uses a `TokenBucket`, implemented using a semaphore, that
+/// limits operations to a specific rate. The token bucket will be refilled gradually.
+/// When the rate is exceeded, the `acquire` method will await until a token is available.
+/// ```
+/// use std::sync::Arc;
+/// use tokio::sync::{AcquireError, Semaphore, SemaphorePermit};
+/// use tokio::time::{sleep, Duration};
+///
+/// struct TokenBucket {
+///     sem: Arc<Semaphore>,
+///     jh: tokio::task::JoinHandle<()>,
+/// }
+///
+/// impl TokenBucket {
+///     fn new(rate: usize) -> Self {
```

Comment:
Perhaps this should just take a Duration of the amount of time there should be between each permit? That would also allow for permits being added more rarely than once per second.